### ISSUE TITLE
pretty python exception traceback in rust

### ISSUE THREAD

**Issue by jmrgibson** [PyO3/pyo3#592]
The `print` method on `PyErr` (https://pyo3.rs/master/doc/pyo3/struct.PyErr.html#method.print)   is super handy but relies on stderr being available on the python side. When embedding rust + python into another application this is not always available and so it fails silently. It would be better if we had a way of getting the string of the error or the traceback as a string in *rust*, so that we can use it with whatever logging/error infrastructure is already set up on the rust side (eg `log!`, the `failure` crate, etc)

At the moment I've been doing some variation of `py.eval("str(err)")` or `py.eval(import traceback; # etc)` which works, but is cumbersome.

**Comment by kngwyu** [comment]
> When embedding rust + python into another application this is not always available and so it fails silently.

Hmm, I haven't thought about such a situation :thinking: 
Thank you for your feedback!

**Comment by flaper87** [comment]
> At the moment I've been doing some variation of py.eval("str(err)") or py.eval(import traceback; # etc) which works, but is cumbersome.

Do you have a full example of how you're extracting the traceback string? I'd like to do the same.

It'd be great to have a way to extract the traceback str representation built-in into `PyErr`

**Comment by hfingler** [comment]
Same here, need to extract the error since stderr is not available :(

**Comment by davidhewitt** [comment]
I've done a lot of reworking of errors for `0.12`, so it would be nice to have an API for this included in this release if possible.

**Comment by sebpuetz** [comment]
Is this issue blocked by other work on `PyErr` or can this be done independently of your planned changes?

**Comment by davidhewitt** [comment]
I think it's easiest to do it once #1115 merges, if implementing display isn't enough. I was wondering about adding a `format_traceback()` function. But not worth doing right now as basically the whole `PyErr` implementation is mid-rewrite.

**Comment by davidhewitt** [comment]
For those who discover this thread: the new reworked `PyErr` now implements `Display`, so it's easy enough to get the error type and value.

That `Display` doesn't currently show the traceback, so as discussed above I'm considering adding a `format_traceback()` function. But I think this function will pretty much call `import traceback; traceback.format_exception()`, so I'm wondering whether we really want this API in the pyo3 core (it's easy for users to re-implement).

So I'd like a little bit more time to think about exactly what that API should be, if we do include it.

**Comment by davidhewitt** [comment]
Alternatively we can use [`PyTraceback_Print`](https://github.com/python/cpython/blob/ccdcb20cbeb1943f506a422c3fd942651389a187/Include/traceback.h#L10) here, but that needs a file object as the second argument. (We'd potentially want to implement a custom file object for performance.)
